src/battle_end_turn.c

```diff
@@ -0,0 +1,1589 @@
+    return TRUE;
+}
+
+static bool32 HandleEndTurnOrder(u32 battler)
+{
+    bool32 effect = FALSE;
+
+    gBattleTurnCounter++;
+    gBattleStruct->endTurnEventsCounter++;
+
+    u32 i, j;
+    for (i = 0; i < gBattlersCount; i++)
+    {
+        gBattlerByTurnOrder[i] = i;
+    }
+    for (i = 0; i < gBattlersCount - 1; i++)
+    {
+        for (j = i + 1; j < gBattlersCount; j++)
+        {
+            if (GetWhichBattlerFaster(gBattlerByTurnOrder[i], gBattlerByTurnOrder[j], FALSE) == -1)
+                SwapTurnOrder(i, j);
+        }
+    }
+
+    return effect;
+}
+
+static bool32 HandleEndTurnVarious(u32 battler)
+{
+    u32 i;
+    bool32 effect = FALSE;
+
+    gBattleStruct->endTurnEventsCounter++;
+
+    if (gFieldStatuses & STATUS_FIELD_FAIRY_LOCK && gFieldTimers.fairyLockTimer == gBattleTurnCounter)
+        gFieldStatuses &= ~STATUS_FIELD_FAIRY_LOCK;
+
+    for (i = 0; i < NUM_BATTLE_SIDES; i++)
+    {
+        if (gSideStatuses[i] & SIDE_STATUS_DAMAGE_NON_TYPES && gSideTimers[i].damageNonTypesTimer == gBattleTurnCounter)
+            gSideStatuses[i] &= ~SIDE_STATUS_DAMAGE_NON_TYPES;
+    }
+
+    for (i = 0; i < gBattlersCount; i++)
+    {
+        if (gStatuses3[i] & STATUS3_ALWAYS_HITS)
+            gStatuses3[i] -= STATUS3_ALWAYS_HITS_TURN(1);
+
+        if (gDisableStructs[i].chargeTimer && --gDisableStructs[i].chargeTimer == 0)
+            gStatuses3[i] &= ~STATUS3_CHARGED_UP;
+
+        if (gStatuses3[i] & STATUS3_LASER_FOCUS && gDisableStructs[i].laserFocusTimer == gBattleTurnCounter)
+            gStatuses3[i] &= ~STATUS3_LASER_FOCUS;
+
+        gBattleStruct->hpBefore[i] = gBattleMons[i].hp;
+    }
+
+    return effect;
+}
+
+static bool32 HandleEndTurnWeather(u32 battler)
+{
+    gBattleStruct->endTurnEventsCounter++;
+    return EndOrContinueWeather();
+}
+
+static bool32 HandleEndTurnWeatherDamage(u32 battler)
+{
+    bool32 effect = FALSE;
+
+    u32 ability = GetBattlerAbility(battler);
+    u32 currBattleWeather = GetCurrentBattleWeather();
+
+    if (currBattleWeather == 0xFF)
+    {
+        // If there is no weather on the field, no need to check other battlers so go to next state
+        gBattleStruct->turnEffectsBattlerId = 0;
+        gBattleStruct->endTurnEventsCounter++;
+        return effect;
+    }
+
+    gBattleStruct->turnEffectsBattlerId++;
+
+    if (!IsBattlerAlive(battler) || !HasWeatherEffect())
+        return effect;
+
+
+    switch (currBattleWeather)
+    {
+    case BATTLE_WEATHER_FOG:
+    case BATTLE_WEATHER_STRONG_WINDS:
+        break;
+    case BATTLE_WEATHER_RAIN:
+    case BATTLE_WEATHER_RAIN_PRIMAL:
+    case BATTLE_WEATHER_RAIN_DOWNPOUR:
+        if (ability == ABILITY_DRY_SKIN || ability == ABILITY_RAIN_DISH)
+        {
+            if (AbilityBattleEffects(ABILITYEFFECT_ENDTURN, battler, ability, 0, MOVE_NONE))
+                effect = TRUE;
+        }
+        break;
+    case BATTLE_WEATHER_SUN:
+    case BATTLE_WEATHER_SUN_PRIMAL:
+        if (ability == ABILITY_DRY_SKIN)
```

Comment:
```suggestion
        if (ability == ABILITY_DRY_SKIN || ability == ABILITY_SOLAR_POWER)
```
According to the smogon research